BJS2-15628

Caveri95 opened the pull request:
- Добавлена возможность создавать компании для сбора бюджета к проектам
- Редактировать/создавать компании может только владелей или менеджер
- Написаны тесты к сервисам

Caveri95 committed:
BJS2-15628

- Добавлена возможность создавать компании для сбора бюджета к проектам
- Редактировать/создавать компании может только владелей или менеджер
- Написаны тесты к сервисам

pablo023 commented on src/main/java/faang/school/projectservice/controller/CampaignController.java:
Респонс статусы не проставил.

Caveri95 replied:
Да, окей, добавлю

Caveri95 replied:
по дефолту и так контроллер 200 вернет, если ошибок не было. Лучше тогда статусы проставлять на ответы, отличные от 200?

pablo023 commented on src/main/java/faang/school/projectservice/dto/campaign/CampaignUpdateDto.java:
```diff
@@ -0,0 +1,28 @@
+    private String description;
+
+    @Min(1)
+    private BigDecimal goal;
```
Почему это бигдецимал?

Caveri95 replied:
Посмотрел какой тип используется при создании в БД, там DECIMAL. Поэтому выбрал аналогичный тут.
Можно конечно и Double использовать

pablo023 commented on src/main/java/faang/school/projectservice/service/CampaignService.java:
```diff
@@ -0,0 +1,105 @@
+        campaignDto.setCreatedBy(userId);
+        campaignDto.setUpdatedBy(userId);
+
+        checkUserPermission(campaignDto.getProjectId(), userId);
```
Я бы это в валидатор вынес.

Caveri95 replied:
Согласен

pablo023 commented on src/main/java/faang/school/projectservice/service/CampaignService.java:
```diff
@@ -0,0 +1,105 @@
+                .flatMap(filter -> filter.apply(campaigns, campaignFiltersDto))
+                .toList());
+
+        if (campaignDtos.isEmpty()) {
+            return List.of();
+        } else {
+            return campaignDtos;
+        }
```
Не особо имеет смысл. Если campaigns будет пустым, мы его вернем пустым, зачем нам отдельный List.of() для этого?

Caveri95 replied:
Ага, что то намудрил

pablo023 commented on src/main/java/faang/school/projectservice/service/ProjectService.java:
```diff
@@ -0,0 +1,31 @@
+                .anyMatch(teamMember -> teamMember.getRoles().contains(TeamRole.MANAGER));
+    }
+
+    public Project getProjectById(Long projectId) {
+        return projectRepository.getProjectById(projectId);
+    }
```
Не надо добавлять проверку на то, существует ли он?

Caveri95 replied:
внутри этого метода есть проверка уже 
![image](https://github.com/user-attachments/assets/0fe1682a-7197-45c0-a9ba-02c13d8e6bb6)

pablo023 reviewed (changes requested):
Вообще, очень даже хорошо, но чуть чуть исправь по замечаниям

Caveri95 committed:
BJS2-15628

- Вынесена логика по проверки прав в отдельный сервис валидатор
- Правки по замечаниям
- Поправлены тесты и дописан тест на класс валидатор

pablo023 commented:
Пофикси конфликты, и можем мерджить!

Caveri95 committed:
Merge branch 'titan-master-bc5' into titan-master-bc5-BJS2-15628

# Conflicts:
#	src/main/java/faang/school/projectservice/controller/GlobalExceptionHandler.java
#	src/main/java/faang/school/projectservice/exception/DataValidationException.java
#	src/main/java/faang/school/projectservice/service/ProjectService.java
#	src/main/java/faang/school/projectservice/validator/ProjectValidator.java
#	src/test/java/faang/school/projectservice/service/ProjectServiceTest.java
#	src/test/java/faang/school/projectservice/validator/ProjectValidatorTest.java

Caveri95 committed:
BJS2-15628

- Решены конфликты
